virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,150 @@ def images_define_by_params(self, name, image_params, media=None,
+        legacy_dev = qdevices.QStringDevice('SER-%s' % serial_id,
+                                            cmdline=legacy_cmd)
+        arm_serial = (serial_type == 'isa-serial'
+                      and 'arm' in serial_params.get("machine_type", ""))
+        if (arm_serial or not self.has_option("chardev")
+            or not self.has_device(serial_type)):
+            return legacy_dev
+
+        virtio_port_type = serial_params.get('virtio_port_type')
+        if virtio_port_type:
+            if "-mmio" in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
+
+        prefix = serial_params.get('virtio_port_name_prefix')
+        name = serial_id
+        name = prefix + name if prefix else name
+        serial_params['name'] = name
+
+        chardev_args = {'logfile': serial_params.get('chardev_logfile'),
+                        'logappend': serial_params.get('chardev_logappend')}
+        backend = serial_params.get('chardev_backend', 'unix_socket')
+        if backend not in ['ringbuf', 'spiceport', 'spicevmc']:
+            chardev_args.update({'mux': serial_params.get('chardev_mux')})
+        if backend in ['unix_socket', 'file', 'pipe', 'serial',
+                       'tty', 'parallel', 'parport']:
+            chardev_args.update({'path': file_name})
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_args.update({'debug': serial_params.get('chardev_debug'),
+                                 'name': name})
+        elif backend == 'stdio':
+            chardev_args.update({'signal': serial_params.get('chardev_signal')})
+        elif backend in ['udp', 'tcp_socket']:
+            chardev_args.update({'host': serial_params.get('chardev_host'),
+                                 'port': serial_params.get('chardev_port'),
+                                 'ipv4': serial_params.get('chardev_ipv4'),
+                                 'ipv6': serial_params.get('chardev_ipv6')})
+            if backend == 'udp':
+                chardev_args.update(
+                    {'localaddr': serial_params.get('chardev_localaddr'),
+                     'localport': serial_params.get('chardev_localport')})
+            else:
+                chardev_args.update({'to': serial_params.get('chardev_to'),
+                                     'nodelay': serial_params.get('chardev_nodelay')})
+        if 'socket' in backend:  # tcp_socket & unix_socket
+            chardev_args.update({'server': serial_params.get('chardev_server', 'yes'),
+                                 'nowait': serial_params.get('chardev_nowait', 'yes'),
+                                 'reconnect': serial_params.get('chardev_reconnect')})
```

Comment:
@vivianQizhu  Checked all the options for `-chardev`.
1. For backend `ringbuf`, miss `size`.
2. For backend `unix_socket`, miss `telnet`.   and miss `websocket` only for qemu-3.1.0.
3. For backend `tcp_socket`, miss `telnet`, `tls-creds`, and miss  `websocket` only for qemu-3.1.0.
4. For backend `vc`, miss `[[,width=width][,height=height]][[,cols=cols][,rows=rows]]`.

I do not know if there are cases test these options, ignore me if no need to add them specially.

Reply:
@yanan-fu `telnet` and `tls-creds` should be included, for others there are no requirements currently. I will update it for `telnet` and `tls-creds` at least. Thanks.

Reply:
@yanan-fu I am not adding `tls-creds` for now, the test case related to it requires complex configuration and in low priority, we have no plan to automate it recently.